token-swap/program/src/curve/stable.rs

```diff
@@ -201,37 +208,27 @@ impl CurveCalculator for StableCurve {
         swap_token_b_amount: u128,
         pool_supply: u128,
         trade_direction: TradeDirection,
-        round_direction: RoundDirection,
+        liquidity_provider_operation: LiquidityProviderOperation,
     ) -> Option<u128> {
         let (swap_source_amount, swap_destination_amount) = match trade_direction {
             TradeDirection::AtoB => (swap_token_a_amount, swap_token_b_amount),
-            TradeDirection::BtoA => (swap_token_b_amount, swap_token_a_amount),
+            TradeDirection::BtoA => (swap_token_b_amount, swap_token_a_amount)
         };
         let leverage = self.amp.checked_mul(N_COINS as u64)?;
         let d0 = compute_d(leverage, swap_source_amount, swap_destination_amount)?;
-        let (new_swap_source_amount, new_swap_destination_amount) = match trade_direction {
-            TradeDirection::AtoB => (swap_token_a_amount + source_amount, swap_token_b_amount),
-            TradeDirection::BtoA => (swap_token_b_amount - source_amount, swap_token_a_amount),
+        let (new_swap_source_amount, new_swap_destination_amount) = match liquidity_provider_operation {
+            LiquidityProviderOperation::Deposit => (swap_token_a_amount + source_amount, swap_token_b_amount),
+            LiquidityProviderOperation::Withdrawal => (swap_token_b_amount - source_amount, swap_token_a_amount)
         };
-        let d1 = compute_d(
-            leverage,
-            new_swap_source_amount,
-            new_swap_destination_amount,
-        )?;
-        let diff = match trade_direction {
-            TradeDirection::AtoB => d1.checked_sub(d0)?,
-            TradeDirection::BtoA => d0.checked_sub(d1)?,
+        let d1 = compute_d(leverage, new_swap_source_amount, new_swap_destination_amount)?;
+        let diff = match liquidity_provider_operation {
+            LiquidityProviderOperation::Deposit => d1.checked_sub(d0)?,
+            LiquidityProviderOperation::Withdrawal => d0.checked_sub(d1)?
         };
         let final_amount = (diff.checked_mul(pool_supply))?.checked_div(d0)?;
-        match round_direction {
-            RoundDirection::Floor => {
-                Some(PreciseNumber::new(final_amount)?.floor()?.to_imprecise()?)
-            }
-            RoundDirection::Ceiling => Some(
-                PreciseNumber::new(final_amount)?
-                    .ceiling()?
-                    .to_imprecise()?,
-            ),
+        match liquidity_provider_operation {
+            LiquidityProviderOperation::Deposit => Some(PreciseNumber::new(final_amount)?.ceiling()?.to_imprecise()?),
+            LiquidityProviderOperation::Withdrawal => Some(PreciseNumber::new(final_amount)?.floor()?.to_imprecise()?)
```

Comment:
`floor()` and `ceiling()` should be flipped here

Reply:
done